Updated Member and StripeCustomerSubscription relations

allouis opened the pull request:
no-issue

This includes changes to move toward using proper relations rather than the custom decoration and listing methods that we currently use for members.

I've updated the serialize method on the models, so that we can deal with models in application code, and when converting to JSON for transport, we retain the same structure that we currently have. This greatly simplifies the changes as there are a lot of places that rely on this structure (specifically the `member.stripe.subscriptions` array and the form of the `subscription` objects in the array)

allouis committed:
Added stripeSubscriptions relation to member model

no-issue

This allows us to fetch the subscriptions for a member via standard
model usage, e.g. `withRelated: ['stripeSubscriptions']` rather than
offloading to loops and `decorateWithSubscriptions` functions, this is
more performant and less non-standard than the existing method.

allouis committed:
Updated serialize methods to match existing format

no-issue

The current usage of `decorateWithSubscriptions` and the usage of
members throughout the codebase has a subscriptions array on a stripe
object on the member, this ensures that when we serialize members to
JSON that we are using the same format.

There is definitely room to change this in future, but this is an
attempt to create as few breaking changes as possible.

allouis commented:
Hey @naz would super appreciate a quick look at this, just wanted to check it makes sense to you and feels like the right direction to moving away from `decorateWithSubscriptions` and such?

naz commented on core/server/models/stripe-customer-subscription.js:
```diff
@@ -1,11 +1,48 @@
 const ghostBookshelf = require('./base');
 
+const CURRENCY_SYMBOLS = {
```
We are almost striking the "rule of three" with this `CURRENCY_SYMBOLS` variable :sweat_smile:  It's been used in [these](https://github.com/TryGhost/Ghost/blob/b61ccf0889c2e813886fed62a7b95a52926c70c2/core/server/services/members/config.js#L48-L55) two [places](https://github.com/TryGhost/Members/blob/395b2048998c9d43fd37a0f48135dbbdec05aaca/packages/members-api/lib/stripe/index.js#L8-L15).

 I'm starting to think if it's low effort maybe worth moving it down to [StripePaymentProcessor](https://github.com/TryGhost/Members/blob/395b2048998c9d43fd37a0f48135dbbdec05aaca/packages/members-api/lib/stripe/index.js#L17) class's fields and keep it there as an exposed constant?

allouis replied:
It's now been removed from the the `StripePaymentProcessor` class completely - and it feels like the model/data layer is a better place for it. I'm going to leave it as is for now, until we can work out the best place for it :thinking:

naz commented on core/server/models/stripe-customer-subscription.js:
```diff
@@ -1,11 +1,48 @@
+            id: defaultSerializedObject.subscription_id,
+            customer: {
+                id: defaultSerializedObject.customer_id,
+                // TODO? The customer is not fetched by default so these sometimes won't exist
```
It used to be fetched by default in memebers-api, right? Don't think we'd be breaking much if we stop loading them by default, although it's technically a breaking change if we do. 

I'm also not sure we allow passing any includes for member's endpoints, so that's something that would have to be taken care if we stop fetching this relation.

allouis replied:
It did yeah!

My thinking was that in the members controllers we would just use `{withRelated: ['stripeSubscriptions', 'stripeSubscriptions.customer']}` in the config - that would keep the same functionality, but would allow for more performant queries when we _don't_ need the subscriptions - what do you think?

naz replied:
Yup, the thinking is straight to the point! The the endpoints are not stable yet, and whole members is beta so I think even if we start returning members without subscriptions by default we should be fine :)

naz replied:
To be extra clear, would be cool to keep the functionality as is - with subscriptions by default, but we could discuss with product changing this to not return them by default which I'd be in favor of :)

allouis replied:
Oh yeah, I totally think the HTTP API should be kept as it is for now! If you look at the `members.js` controller file, you'll see I've updated all the methods to use `withRelated` :) 

> but we could discuss with product changing this to not return them by default which I'd be in favor of :)

Yeah I _think_ I would to

naz commented:
It looks like a good direction to me :+1:  Think something we'd def. wanna change with `decorateWithSubscriptions` removal is introduce `includes` options in `members` controller to keep the same convention as with other resources

allouis committed:
fixup! Added stripeSubscriptions relation to member model

allouis committed:
Used model methods in Admin API members controller

no-issue

These methods use bookshelf relations to 1) remove a loop when listing
members and 2) fetch the subscriptions using relations which is more
performant

allouis committed:
Added filter to stripeSubscriptions relation

no-issue

This matches previous functionality of only every returning the active
subscriptions, it may be worth considering which subscriptions we store
in future, so we can possible remove this filter

allouis committed:
fixup! Used model methods in Admin API members controller

allouis committed:
Replaced all usage of member models with members-api

naz commented on core/server/api/canary/members.js:
```diff
@@ -14,18 +14,6 @@ const logging = require('../../../shared/logging');
 const _ = require('lodash');
 
-const decorateWithSubscriptions = async function (member) {
-    // NOTE: this logic is here until relations between Members/MemberStripeCustomer/StripeCustomerSubscription
```
omg finally :fire:

naz commented on core/server/api/canary/members.js:
```diff
@@ -761,7 +761,7 @@ module.exports = {
                                 await Promise.map(membersWithStripeCustomers, async (stripeMember) => {
                                     try {
-                                        await membersService.api.members.linkStripeCustomer(stripeMember.stripe_customer_id, stripeMember);
+                                        await membersService.api.members.linkStripeCustomerById(stripeMember.stripe_customer_id, stripeMember.id);
```
Something I don't have strong opinion about but wanted to know what you think. Did you consider  keeping existing `linkStripeCustomer` methods as is without renaming to `ById`? Instead check for type of second argument that's passed in - if integer would fetch the records, if model do the normal logic.

I think the `ById` approach is very explicit and gives a signal about possible things that might be going on inside the method call, like fetching the model. On the other side, it expands the API surface :thinking:

allouis replied:
I did consider that initially yeah, the reason I decided against it is twofold:
1) Like you say this is more explicit, and I'm finding as this codebase grows, explicitness is becoming more valuable
2) I'm hoping to "type" the `members-api` code in the future, and JSDoc typings with polymorphic functions is quite tricky!

allouis committed:
Installed @tryghost/members-api@0.26.0
